### ISSUE TITLE
What primitives are safe to share across tasks?

### ISSUE THREAD

**Issue by alshain** [ronf/asyncssh#119]
Can I share a single connection across many coroutines?

It would be nice to document what can and what can't be shared across different coroutines.

From what I understand, it's possible that while an `asyncssh` coroutine is executing, the related asyncssh object may be in an inconsistent state, so starting another asyncssh coroutine asynchronously might break things.

**Comment by ronf** [comment]
Yes - connections can be shared without a problem. Given that a connection can contain multiple channels, this is actually a very common use case, and AsyncSSH is designed to allow this. One of the great things about coroutines vs. threads is that you can control when a coroutine gives up control, allowing you to make sure that objects are _not_ in an inconsistent state when a coroutine yields control to others.

In fact, you can even have multiple coroutines sharing a single channel. A good example of that is the SFTP code in AsyncSSH which lets you have multiple file handles open on a single SFTP client session and do I/O on those file handles from multiple coroutines in parallel. The SFTP "file copier" function even does multiple writes in parallel to the _same_ file handle by creating separate tasks (coroutines) for each write call (and each write has an explicit start position in it to avoid ordering concerns), and that all works fine.

There are some things to keep in mind if you want to do this yourself, but there isn't much to it:

- The main thing that's explicitly not allowed is that you can't have two coroutines attempting to read from the same StreamReader at the same time. It's ok to do this on streams associated with different channels sharing a connection, but not on the same channel, as it wouldn't be predictable which reader got which data. AsyncSSH explicitly raises an error if you try this. Instead, you should create a single coroutine to read data from the stream and then figure out yourself how to get it to the right other coroutines that are waiting for it.

- When you do writes, they will be "atomic", meaning all of the data you write will be guaranteed to be consecutive no matter what other coroutines might be writing to the same channel. However, if you have a message that you break up into multiple pieces and call write() separately on each piece, you need to make sure you don't do any blocking operation that might cause your coroutine to yield in between these writes if you want that guarantee. For instance, if you mix read & write calls, you can't guarantee that another coroutine won't write to the channel in between while you are blocked on a read.

That's about it. If you see any cases where simultaneous access of this sort doesn't work the way you expect, please let me know!

**Comment by alshain** [comment]
👍 Thanks a lot for the information. Awesomely detailed as always.

> The main thing that's explicitly not allowed is that you can't have two coroutines attempting to read from the same StreamReader at the same time. It's ok to do this on streams associated with different channels sharing a connection, but not on the same channel, as it wouldn't be predictable which reader got which data. AsyncSSH explicitly raises an error if you try this.

Good to know. I think I did something like this. I had a coroutine awaiting a line from stderr, and another coroutine doing `process.communicate()`, but I never saw an error from that. I rewrote that part though :)

I was running into #55 , which prompted this question, because I needed to know how easily I could reuse one connection to a single server.

Thanks again!

**Comment by ronf** [comment]
Mixing SSHProcess communicate() and SSHReader read calls is an interesting edge case. You're right that AsyncSSH doesn't actively prevent that right now. The communicate() call only collects output left in the receive buffer after wait_closed() on the channel, though, while any readers waiting on data will be unblocked before that when EOF is processed on the channel. So, I think the readers will get first crack at the data, and whatever isn't read by them will be delivered in the data returned by communicate().

**Comment by ronf** [comment]
Closing this for now, but feel free to open a new issue if you have additional questions or concerns.
